src/devices/rolling_code.c

```diff
@@ -0,0 +1,335 @@
+        uint8_t nibble = bitrow_get_bit(bit_buffer, (*bit_index)++) << 3 | 
+            bitrow_get_bit(bit_buffer, (*bit_index)++) << 2 | 
+            bitrow_get_bit(bit_buffer, (*bit_index)++) << 1 | 
+            bitrow_get_bit(bit_buffer, (*bit_index)++);
+        
+        if (nibble == 0x01) nibble_buffer[nibble_index++] = 0;
+        else if (nibble == 0x03) nibble_buffer[nibble_index++] = 1;
+        else if (nibble == 0x07) nibble_buffer[nibble_index++] = 2;
+        else {
+            fprintf(stderr, "Unknown nibble %02x\n", nibble);
+            return 1;
+        }
+    }
+
+    if (nibble_index != TRINARY_SIZE) {
+        fprintf(stderr, "Not enough bits for %d nibbles\n", TRINARY_SIZE);
+
+        return 1;
+    } else {
+        return 0;
+    }
+}
+
+static void fix_a_code(uint8_t *a_raw, uint8_t *r_raw, uint8_t *dst) {
+    uint8_t prev = 0;
+    for (int i = 0; i < RAW_SIZE; i++) {
+        int16_t x = a_raw[i] - r_raw[i];
+        if (x < 0) x += 3;
+        x = x - prev;
+        if (x < 0) x += 3;
+        dst[i] = x & 0xff;
+        prev = a_raw[i];
+    }
+}
+
+// Buffer has to be len + 1 bytes long.
+static void raw_to_chars(uint8_t *src, char *dst, int len) {
+    for (int i = 0; i < len; i++) {
+        dst[i] = src[i] + '0';
+    }
+
+    dst[len] = '\0';
+}
+
+static uint32_t raw_to_uint(uint8_t *ptr, int len) {
+    uint32_t value = 0;
+    
+    for (int i = 0; i < len; i++) {
+        value *= 3;
+        value += ptr[i];
+    }
+
+    return value;
+}
+
+static void rotate_mirror_add(uint32_t *counter, uint32_t *mirror, int c) {
+    uint32_t result = 0;
+    for (int i = 0; i < c; i++) {
+        uint8_t msb = *mirror >> 31;
+        *mirror <<= 1;
+        *mirror += msb;
+    }
+
+    *counter += *mirror;
+}
+
+static void mirror_counter(uint32_t *counter, uint32_t *mirror) {
+    uint32_t result = 0;
+    for (int i = 0; i < 32; i++) {
+        uint8_t bit = *counter & 0x01;
+        *counter >>= 1;
+        result <<= 1;
+        result |= bit;
+    }
+
+    *mirror = result;
+}
+
+static uint32_t get_rolling_code(uint32_t r1, uint32_t r3) {
+    uint32_t counter = r3;
+    uint32_t mirror = r1;
+
+    counter += mirror;
+
+    rotate_mirror_add(&counter, &mirror, 3);
+    rotate_mirror_add(&counter, &mirror, 2);
+    rotate_mirror_add(&counter, &mirror, 2);
+    rotate_mirror_add(&counter, &mirror, 2);
+    rotate_mirror_add(&counter, &mirror, 1);
+    rotate_mirror_add(&counter, &mirror, 3);
+    rotate_mirror_add(&counter, &mirror, 1);
+    rotate_mirror_add(&counter, &mirror, 1);
+
+    mirror_counter(&counter, &mirror);
+
+    return mirror;
+}
+
+static int rolling_code_decode(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+    // This holds previous start width 1 fixed and rolling codes for 
+    // counter decoding once we have a start width 3 packet.
+    static uint8_t prev_1_a_corrected[RAW_SIZE] = {NOT_SET};
+    static uint8_t prev_1_r_raw[RAW_SIZE] = {NOT_SET};
```

Comment:
Decoders usually doesn't hold any state. And I don't understand the bit stream enough to say if it really is needed. Can you add some signal samples to the tests repository?

Reply:
To garner any useful data from the transmission it is unfortunately necessary to retain a small amount of state. I'd hoped that all of the data bursts could be captured in a single bitbuffer with several rows, but the delay between bursts runs afoul of the limits set forth in pulse_decode.c. Christian and I discussed this briefly in email - here's why, without changes in pulse_decode and other files, that state is needed:

1) AFAIK RTL_433 ... can't handle trinary. That means I have to use the OOK_PCM_RZ format with the short and long widths set to 500 uS (the smallest common value). This is not a big deal, except for #2 and #3.
2) When I attempt to grab the data, the 60 mS gap between the packets triggers the End-Of-Packet processing on lines 349-ish in pulse_decode.c. It seems that I cannot ever get the packets _either_ on one row or on their own row!
3) Even if we fix #2, OOK_PCM_RZ code doesn't support the gap_limit, so I would get all of the values on a single row. 

Note that I did provide for debugging output - if you use -vvvv, you get output that has the raw values in each of the packets received.

I will gladly provide sample data files once my new transmitters arrive tomorrow. I'm not inclined to post files with my actual data in them.
 
Below are examples of the trinary nature of the data and an example of the 60 mS gap between the '1' start-bit-width packet and the '3' start-bit-width packet.

<img width="826" alt="RCR Gap Edited" src="https://user-images.githubusercontent.com/3858971/77342790-1f90bf80-6d07-11ea-850d-0ee149fb87c6.png">
<img width="531" alt="RCR Data" src="https://user-images.githubusercontent.com/3858971/77342798-23244680-6d07-11ea-90ab-0ea5884f87d3.png">

Reply:
Ok, lets call for help. @zuckschwerdt I looked at the pulse_demod_pcm() code and it seemed feasible to add gap_limit support. Do you concur?

Somewhere around this line:
`periods = MIN(periods, max_zeros);
`
there could be some logic to add rows.

Reply:
I didn't even realize that the PCM slicer is the problem here. Sure, I wanted to add that anyway.

Reply:
@zuckschwerdt in pulse_detect:371 this condition seems strange:

```
                if (((s->pulse_length > (PD_MAX_GAP_RATIO * s->max_pulse))    // gap/pulse ratio exceeded
                        && (s->pulse_length > (PD_MIN_GAP_MS * samples_per_ms)))    // Minimum gap exceeded
                        || (s->pulse_length > (PD_MAX_GAP_MS * samples_per_ms))) {    // maximum gap exceeded
```

I think this also will hinder the merging of signals with long gaps. And raising the reset limit to >60ms will reduce the time resolution with regards to different signals. I am not sure what it is now but I think it is like 20ms.

It would be nice to know these metrics. For example FSK devices doesn't have this property but are still affected by it.

As always I'm not totally sure what to do :) it is becoming difficult to decode all signals.

Reply:
It's: below 10ms will never EOP above 100ms will always EOP, between that at longest pulse x10 will EOP. That way we currently can't have gaps longer than "longest pulse x 10" (if that's above 10ms). That's for the AM envelope, thus FM isn't limited by this?